WriteBlobWriteParams could not be found (even doe the js was loaded)

I was trying to adjust the WriteAsync functions a little (adding a write for a DotNetStreamReference) and stumbled over the fact, that my method added to the js file wasn't found. Same goes for the existing version using "WriteBlobWriteParams".

Am I doing something wrong? Was it tested, cause there is no example using the above. 
Could it reference a different IJSObjectReference?

The file is loaded as of what I see in the Devtools sources tab, but I receive an error msg:

Microsoft.JSInterop.JSException: Could not find 'WriteBlobWriteParams' ('WriteBlobWriteParams' was undefined).
blazor.webassembly.js:1 Error: Could not find 'WriteBlobWriteParams' ('WriteBlobWriteParams' was undefined).

Thanks for your reply

--- a/src/KristofferStrube.Blazor.FileSystem/Converters/BlobConverter.cs
+++ b/src/KristofferStrube.Blazor.FileSystem/Converters/BlobConverter.cs
@@ -1,0 +1,18 @@
+﻿using KristofferStrube.Blazor.FileAPI;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace KristofferStrube.Blazor.FileSystem.Converters;
+
+internal class BlobConverter : JsonConverter<Blob>
+{
+    public override Blob Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        throw new NotSupportedException("Does not support deserializing Blobs");
+    }
+
+    public override void Write(Utf8JsonWriter writer, Blob value, JsonSerializerOptions options)
+    {
+        JsonSerializer.Serialize(writer, value.JSReference, options);
+    }
+}

--- a/src/KristofferStrube.Blazor.FileSystem/FileSystemFileHandle.cs
+++ b/src/KristofferStrube.Blazor.FileSystem/FileSystemFileHandle.cs
@@ -28,6 +28,6 @@ public class FileSystemFileHandle : FileSystemHandle
     public async Task<FileSystemWritableFileStream> CreateWritableAsync(FileSystemCreateWritableOptions? fileSystemCreateWritableOptions = null)
     {
         IJSObjectReference jSFileSystemWritableFileStream = await JSReference.InvokeAsync<IJSObjectReference>("createWritable", fileSystemCreateWritableOptions);
-        return new FileSystemWritableFileStream(jSRuntime, jSFileSystemWritableFileStream);
+        return await FileSystemWritableFileStream.CreateAsync(jSRuntime, jSFileSystemWritableFileStream);
     }
 }

--- a/src/KristofferStrube.Blazor.FileSystem/FileSystemWritableFileStream.cs
+++ b/src/KristofferStrube.Blazor.FileSystem/FileSystemWritableFileStream.cs
@@ -1,4 +1,5 @@
 ﻿using KristofferStrube.Blazor.FileAPI;
+using KristofferStrube.Blazor.FileSystem.Extensions;
 using KristofferStrube.Blazor.Streams;
 using Microsoft.JSInterop;
 
@@ -9,16 +10,30 @@ namespace KristofferStrube.Blazor.FileSystem;
 /// </summary>
 public class FileSystemWritableFileStream : WritableStream
 {
+    /// <summary>
+    /// A lazily evaluated task that gives access to helper methods for the File System API.
+    /// </summary>
+    protected readonly Lazy<Task<IJSObjectReference>> fileSystemHelperTask;
+
     public override bool CanSeek => true;
 
     public override long Position { get; set; }
 
+    [Obsolete("This will be removed in the next major release as all creator methods should be asynchronous for uniformity. Use CreateAsync instead.")]
     public static new FileSystemWritableFileStream Create(IJSRuntime jSRuntime, IJSObjectReference jSReference)
     {
         return new FileSystemWritableFileStream(jSRuntime, jSReference);
     }
 
-    internal FileSystemWritableFileStream(IJSRuntime jSRuntime, IJSObjectReference jSReference) : base(jSRuntime, jSReference) { }
+    public static new Task<FileSystemWritableFileStream> CreateAsync(IJSRuntime jSRuntime, IJSObjectReference jSReference)
+    {
+        return Task.FromResult(new FileSystemWritableFileStream(jSRuntime, jSReference));
+    }
+
+    protected FileSystemWritableFileStream(IJSRuntime jSRuntime, IJSObjectReference jSReference) : base(jSRuntime, jSReference)
+    {
+        fileSystemHelperTask = new(() => jSRuntime.GetHelperAsync(FileSystemOptions.DefaultInstance));
+    }
 
     public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
     {
@@ -37,13 +52,12 @@ public class FileSystemWritableFileStream : WritableStream
 
     public async Task WriteAsync(Blob data)
     {
-        await JSReference.InvokeVoidAsync("write", data.JSReference);
+        await JSReference.InvokeVoidAsync("write", data);
     }
 
     public async Task WriteAsync(BlobWriteParams data)
     {
-        IJSObjectReference helper = await helperTask.Value;
-        await helper.InvokeVoidAsync("WriteBlobWriteParams", JSReference, data, data.Data?.JSReference);
+        await JSReference.InvokeVoidAsync("write", data);
     }
 
     public async Task WriteAsync(StringWriteParams data)

--- a/src/KristofferStrube.Blazor.FileSystem/FileSystemWritableFileStream.cs
+++ b/src/KristofferStrube.Blazor.FileSystem/FileSystemWritableFileStream.cs
@@ -52,7 +52,7 @@ public class FileSystemWritableFileStream : WritableStream
 
     public async Task WriteAsync(Blob data)
     {
-        await JSReference.InvokeVoidAsync("write", data);
+        await JSReference.InvokeVoidAsync("write", data.JSReference);
     }
 
     public async Task WriteAsync(BlobWriteParams data)

--- a/src/KristofferStrube.Blazor.FileSystem/Options/WriteParams.cs
+++ b/src/KristofferStrube.Blazor.FileSystem/Options/WriteParams.cs
@@ -1,4 +1,5 @@
 ﻿using KristofferStrube.Blazor.FileAPI;
+using KristofferStrube.Blazor.FileSystem.Converters;
 using System.Text.Json.Serialization;
 
 namespace KristofferStrube.Blazor.FileSystem;
@@ -10,7 +11,8 @@ public class BlobWriteParams : BaseWriteParams
         Type = type;
     }
 
-    [JsonIgnore]
+    [JsonPropertyName("data")]
+    [JsonConverter(typeof(BlobConverter))]
     public Blob? Data { get; set; }
 }
 

--- a/src/KristofferStrube.Blazor.FileSystem/wwwroot/KristofferStrube.Blazor.FileSystem.js
+++ b/src/KristofferStrube.Blazor.FileSystem/wwwroot/KristofferStrube.Blazor.FileSystem.js
@@ -10,11 +10,6 @@ export async function arrayFrom(values) {
     return res;
 }
 
-export function WriteBlobWriteParams(fileSystemWritableFileStream, writeParams, blob) {
-    writeParams.data = blob;
-    fileSystemWritableFileStream.write(writeParams);
-}
-
 export async function arrayBuffer(blob) {
     var buffer = await blob.arrayBuffer();
     var bytes = new Uint8Array(buffer);

